README.md
# irb-notebook

⚠️ Work-in-progress starter code for custom notebook renderers in VS Code. Expect this to change as notebooks matures. ⚠️

This starter includes:

 - 🖥️ TypeScript code to create a simple `NotebookOutputRenderer`
 - 📦 A Webpack build for renderer client code
 - ⚡ Support for hot module reloading and safe boilerplate
 - 🎨 CSS modules support

### Running this Sample

 1. `code-insiders irb-notebook`: Open the folder in VS Code Insiders
 1. Hit `F5` to build+debug

### Structure

A Notebook Renderer consists of code that runs in the VS Code Extension Host (Node.js), which registers the renderer and passes data into the UI code running inside a WebView (Browser/DOM).

This uses TypeScript project references. There are three projects in the `src` directory:

 - `extension` contains the code running in Node.js extension host. It's compiled with `tsc`.
 - `client` is the UI code, built by Webpack, with access to the DOM.
 - `common` contains code shared between the extension and client.

When you run `watch`, `compile`, or `dev`, we invoke both `tsc` and `webpack` to compile the extension and the client portion of the code.


CHANGELOG.md
# Change Log

All notable changes to the "irb-notebook" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

webpack.config.js
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const { DefinePlugin } = require('webpack');
const path = require('path');

const makeConfig = (argv, { entry, out, target, library = 'commonjs' }) => ({
    mode: argv.mode,
    devtool: argv.mode === 'production' ? false : 'inline-source-map',
    entry,
    target,
    output: {
        path: path.join(__dirname, path.dirname(out)),
        filename: path.basename(out),
        publicPath: '',
        libraryTarget: library,
        chunkFormat: library,
    },
    externals: {
        vscode: 'commonjs vscode',
        path: 'path',
        http: 'http',
        url: 'url',
        buffer: 'buffer',
        util: 'util'
    },
    resolve: {
        extensions: ['.ts', '.tsx', '.js', '.jsx', '.css'],
        fallback: {}
    },
    experiments: {
        outputModule: true,
    },
    module: {
        rules: [
            // Allow importing ts(x) files:
            {
                test: /\.tsx?$/,
                loader: 'ts-loader',
                options: {
                    configFile: path.join(path.dirname(entry), 'tsconfig.json'),
                    // transpileOnly enables hot-module-replacement
                    transpileOnly: true,
                    compilerOptions: {
                        // Overwrite the noEmit from the client's tsconfig
                        noEmit: false,
                    },
                },
            },
            // Allow importing CSS modules:
            {
                test: /\.css$/,
                use: [
                    'style-loader',
                    {
                        loader: 'css-loader',
                        options: {
                            importLoaders: 1,
                            modules: true,
                        },
                    },
                ],
            },
        ],
    },
    plugins: [
        new ForkTsCheckerWebpackPlugin({
            typescript: {
                configFile: path.join(path.dirname(entry), 'tsconfig.json'),
            },
        }),
        new DefinePlugin({
            // Path from the output filename to the output directory
            __webpack_relative_entrypoint_to_root__: JSON.stringify(
                path.posix.relative(path.posix.dirname(`/index.js`), '/'),
            ),
            scriptUrl: 'import.meta.url',
        }),
    ],
    infrastructureLogging: {
        level: "log", // enables logging required for problem matchers
    },
});

module.exports = (env, argv) => [
    makeConfig(argv, { entry: './src/client/index.ts', out: './out/client/index.js', target: 'web', library: 'module' }),
    makeConfig(argv, { entry: './src/extension/extension.ts', out: './out/extension/extension.js', target: 'node' }),
    makeConfig(argv, { entry: './src/extension/extension.ts', out: './out/extension/extension.web.js', target: 'webworker' }),
];


src/extension/notebook.ts
// import { TextDecoder, TextEncoder } from 'util';
// import * as vscode from 'vscode';
// import * as http from 'http';
// import { DebugProtocol } from '@vscode/debugprotocol';
// import * as path from 'path';

// interface RawNotebookCell {
//   language: string;
//   value: string;
//   kind: vscode.NotebookCellKind;
//   editable?: boolean;
// }

// export class RdbgNotebookSerializer implements vscode.NotebookSerializer {
//   deserializeNotebook(content: Uint8Array, _token: vscode.CancellationToken): vscode.NotebookData | Thenable<vscode.NotebookData> {
//     const decoded = new TextDecoder().decode(content);
//     let rawCell: RawNotebookCell[];
//     try {
//       rawCell = JSON.parse(decoded);
//     } catch (error) {
//       rawCell = [];
//     }

//     const cells = rawCell.map(cell => new vscode.NotebookCellData(cell.kind, cell.value, cell.language));
//     return new vscode.NotebookData(cells);
//   }

//   serializeNotebook(data: vscode.NotebookData, _token: vscode.CancellationToken): Uint8Array | Thenable<Uint8Array> {
//     const rawCell = data.cells.map(cell => {
//       return {
//         language: cell.languageId,
//         value: cell.value,
//         kind: cell.kind,
//       } as RawNotebookCell;
//     });
//     return new TextEncoder().encode(JSON.stringify(rawCell));
//   }
// }

// export class RdbgNotebookController {
//   readonly controllerId = 'irb-notebook-controller';
//   readonly notebookType = 'irb-notebook';
//   readonly label = 'irb Notebook';
//   readonly supportedLanguage = ['ruby'];

//   private readonly _controller: vscode.NotebookController;
//   private _executionOrder = 0;

//   constructor() {
//     this._controller = vscode.notebooks.createNotebookController(
//       this.controllerId,
//       this.notebookType,
//       this.label
//     );

//     this._controller.supportedLanguages = this.supportedLanguage;
//     this._controller.supportsExecutionOrder = true;
//     this._controller.executeHandler = this._execute.bind(this);
//   }

//   private _execute(cells: vscode.NotebookCell[], _notebook: vscode.NotebookDocument, _controller: vscode.NotebookController) {
//     for (let cell of cells) {
//       this._doExecution(cell);
//     }
//   }

//   dispose() {
//     this._controller.dispose();
//   }

//   private async _doExecution(cell: vscode.NotebookCell): Promise<void> {
//     const execution = this._controller.createNotebookCellExecution(cell);
//     execution.executionOrder = ++this._executionOrder;
//     execution.start(Date.now());
//     const req = { expression: cell.document.getText() };
//     const res = await postJSON(req);
//     console.log(res);
//     execution.replaceOutput([
//       new vscode.NotebookCellOutput([
//         vscode.NotebookCellOutputItem.text(JSON.stringify(res.result)),
//         // vscode.NotebookCellOutputItem.json(res.result)
//       ])
//     ]);
//     execution.end(true, Date.now());
//   }
// }

// function postJSON(chunk: any, path = '/'): Promise<any> {
//   const opts: http.RequestOptions = {
//     host: '127.0.0.1',
//     port: 8080,
//     method: 'POST',
//     path,
//     headers: {
//       'Content-Type': 'application/json',
//     }
//   };
//   return new Promise((resolve, reject) => {
//     const clientReq = http.request(opts, (res) => {
//       if (res.statusCode != 200) {
//       }
//       res.setEncoding('utf8');
//       let rawData = '';
//       res.on('data', (chunk: string) => {
//         rawData += chunk;
//       });
//       res.on('end', () => {
//         const obj = JSON.parse(rawData);
//         resolve(obj);
//       });
//     });
//     clientReq.on('error', (e) => {
//     });
//     const json = JSON.stringify(chunk);
//     clientReq.write(json);
//     clientReq.end();
//   });
// }

// export async function debugCell(cell: vscode.NotebookCell, cellToFileMap: Map<string, string>, fileToCellMap: Map<string, vscode.NotebookCell>) {
//   const path = await createFile();
//   await debug(path);

//   async function createFile() {
//     const text = `require "debug/session";DEBUGGER__.open_tcp(port: 12345)\n` + cell.document.getText();
//     const body = { text };
//     const res = await postJSON(body, '/files');
//     cellToFileMap.set(cell.document.uri.toString(), res.path);
//     fileToCellMap.set(res.path, cell);
//     return res.path;
//   }

//   async function debug(path: string) {
//     const res = await postJSON({ path }, '/debug');
//     const sleep = (time: number) => {
//       return new Promise<void>((resolve, reject) => {
//         setTimeout(() => {
//           resolve();
//         }, time);
//       });
//     };
//     await sleep(3000);
//     const c = {
//       type: "rdbg",
//       name: "Attach with rdbg (tcp 12345)",
//       request: "attach",
//       debugPort: "localhost:12345",
//       internalConsoleOptions: false,
//     };
//     vscode.debug.startDebugging(undefined, c);
//   }
// }

// export function hookWillSendMessages(message: any) {

// }

// export class RdbgDebugAdapterTrackerFactory implements vscode.DebugAdapterTrackerFactory {
//   constructor(private readonly cellToFileMap: Map<string, string>, private readonly fileToCellMap: Map<string, vscode.NotebookCell>) { }
//   createDebugAdapterTracker(session: vscode.DebugSession): vscode.ProviderResult<vscode.DebugAdapterTracker> {
//     return {
//       onWillReceiveMessage: (message: any) => {
//         hookSource(message, (source) => {
//           if (source && source.path) {
//             const file = this.cellToFileMap.get(source.path);
//             if (file) {
//               source.path = file;
//             }
//           }
//         });
//       },
//       onDidSendMessage: (message: any) => {
//         hookSource(message, (source) => {
//           if (source && source.path && source.name) {
//             const cell = this.fileToCellMap.get(source.path);
//             if (cell) {
//               source.path = cell.document.uri.toString();
//               source.name = path.basename(cell.document.uri.path);
//             }
//           }
//         });
//       }
//     };
//   }
// }

// function hookSource(message: DebugProtocol.ProtocolMessage, hook: (source: DebugProtocol.Source | undefined) => void) {
//   switch (message.type) {
//     case "request":
//       const req = message as DebugProtocol.Request;
//       switch (req.command) {
//         case "setBreakpoints":
//           const sbr = req as DebugProtocol.SetBreakpointsRequest;
//           hook(sbr.arguments.source);
//           break;
//         case "source":
//           const sr = req as DebugProtocol.SourceRequest;
//           hook(sr.arguments.source);
//           break;
//       }
//       break;
//     case "event":
//       const evt = message as DebugProtocol.Event;
//       switch (evt.event) {
//         case "loadedSource":
//           const lse = evt as DebugProtocol.LoadedSourceEvent;
//           hook(lse.body.source);
//           break;
//       }
//       break;
//     case "response":
//       const res = message as DebugProtocol.Response;
//       switch (res.command) {
//         case "stackTrace":
//           const str = res as DebugProtocol.StackTraceResponse;
//           str.body.stackFrames.forEach((frame) => {
//             hook(frame.source);
//           });
//           break;
//       }
//       break;
//   }
// }


src/extension/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { TextDecoder, TextEncoder } from 'util';
import * as http from 'http';
import { DebugProtocol } from '@vscode/debugprotocol';
import * as path from 'path';

// This method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  const cellToFileMap = new Map<string, string>();
  const fileToCellMap = new Map<string, vscode.NotebookCell>();
  context.subscriptions.push(
    vscode.workspace.registerNotebookSerializer('irb-notebook', new RdbgNotebookSerializer()),

    new RdbgNotebookController(),

    vscode.commands.registerCommand('rdbg.debugCell', (cell: vscode.NotebookCell) => {
      debugCell(cell, cellToFileMap, fileToCellMap);
    }),

    vscode.debug.registerDebugAdapterTrackerFactory('rdbg', new RdbgDebugAdapterTrackerFactory(cellToFileMap, fileToCellMap))
  );
}

// This method is called when your extension is deactivated
export function deactivate() { }


interface RawNotebookCell {
  language: string;
  value: string;
  kind: vscode.NotebookCellKind;
  editable?: boolean;
}

class RdbgNotebookSerializer implements vscode.NotebookSerializer {
  deserializeNotebook(content: Uint8Array, _token: vscode.CancellationToken): vscode.NotebookData | Thenable<vscode.NotebookData> {
    const decoded = new TextDecoder().decode(content);
    let rawCell: RawNotebookCell[];
    try {
      rawCell = JSON.parse(decoded);
    } catch (error) {
      rawCell = [];
    }

    const cells = rawCell.map(cell => new vscode.NotebookCellData(cell.kind, cell.value, cell.language));
    return new vscode.NotebookData(cells);
  }

  serializeNotebook(data: vscode.NotebookData, _token: vscode.CancellationToken): Uint8Array | Thenable<Uint8Array> {
    const rawCell = data.cells.map(cell => {
      return {
        language: cell.languageId,
        value: cell.value,
        kind: cell.kind,
      } as RawNotebookCell;
    });
    return new TextEncoder().encode(JSON.stringify(rawCell));
  }
}

class RdbgNotebookController {
  readonly controllerId = 'irb-notebook-controller';
  readonly notebookType = 'irb-notebook';
  readonly label = 'irb Notebook';
  readonly supportedLanguage = ['ruby'];

  private readonly _controller: vscode.NotebookController;
  private _executionOrder = 0;

  constructor() {
    this._controller = vscode.notebooks.createNotebookController(
      this.controllerId,
      this.notebookType,
      this.label
    );

    this._controller.supportedLanguages = this.supportedLanguage;
    this._controller.supportsExecutionOrder = true;
    this._controller.executeHandler = this._execute.bind(this);
  }

  private _execute(cells: vscode.NotebookCell[], _notebook: vscode.NotebookDocument, _controller: vscode.NotebookController) {
    for (let cell of cells) {
      this._doExecution(cell);
    }
  }

  dispose() {
    this._controller.dispose();
  }

  private async _doExecution(cell: vscode.NotebookCell): Promise<void> {
    const execution = this._controller.createNotebookCellExecution(cell);
    execution.executionOrder = ++this._executionOrder;
    execution.start(Date.now());
    const req = { expression: cell.document.getText() };
    const res = await postJSON(req);
    console.log(res);
    execution.replaceOutput([
      new vscode.NotebookCellOutput([
        vscode.NotebookCellOutputItem.text(JSON.stringify(res.result)),
        vscode.NotebookCellOutputItem.json(res.result)
      ])
    ]);
    execution.end(true, Date.now());
  }
}

function postJSON(chunk: any, path = '/'): Promise<any> {
  const opts: http.RequestOptions = {
    host: '127.0.0.1',
    port: 8080,
    method: 'POST',
    path,
    headers: {
      'Content-Type': 'application/json',
    }
  };
  return new Promise((resolve, reject) => {
    const clientReq = http.request(opts, (res) => {
      if (res.statusCode !== 200) {
        console.error('not expected code')
      }
      res.setEncoding('utf8');
      let rawData = '';
      res.on('data', (chunk: string) => {
        rawData += chunk;
      });
      res.on('end', () => {
        const obj = JSON.parse(rawData);
        resolve(obj);
      });
    });
    clientReq.on('error', (e) => {
      console.error(e)
    });
    const json = JSON.stringify(chunk);
    clientReq.write(json);
    clientReq.end();
  });
}

async function debugCell(cell: vscode.NotebookCell, cellToFileMap: Map<string, string>, fileToCellMap: Map<string, vscode.NotebookCell>) {
  const path = await createFile();
  await debug(path);

  async function createFile() {
    const text = `require_relative "/Users/s15236/workspace/debug/lib/debug/session";DEBUGGER__.open_tcp(port: 12345)\n` + cell.document.getText();
    const body = { text };
    const res = await postJSON(body, '/files');
    cellToFileMap.set(cell.document.uri.toString(), res.path);
    fileToCellMap.set(res.path, cell);
    fileToCellMap.set(`/private${res.path}`, cell);
    return res.path;
  }

  async function debug(path: string) {
    const res = await postJSON({ path }, '/debug');
    const sleep = (time: number) => {
      return new Promise<void>((resolve, reject) => {
        setTimeout(() => {
          resolve();
        }, time);
      });
    };
    await sleep(3000);
    const c = {
      type: "rdbg",
      name: "Attach with rdbg (tcp 12345)",
      request: "attach",
      debugPort: "localhost:12345",
      internalConsoleOptions: false,
    };
    vscode.debug.startDebugging(undefined, c);
  }
}

class RdbgDebugAdapterTrackerFactory implements vscode.DebugAdapterTrackerFactory {
  constructor(private readonly cellToFileMap: Map<string, string>, private readonly fileToCellMap: Map<string, vscode.NotebookCell>) { }
  createDebugAdapterTracker(session: vscode.DebugSession): vscode.ProviderResult<vscode.DebugAdapterTracker> {
    return {
      onWillReceiveMessage: (message: any) => {
        hookSource(message, (source) => {
          if (source && source.path) {
            const file = this.cellToFileMap.get(source.path);
            if (file) {
              source.path = file;
            }
          }
        });
      },
      onDidSendMessage: (message: any) => {
        hookSource(message, (source) => {
          if (source && source.path && source.name) {
            const cell = this.fileToCellMap.get(source.path);
            if (cell) {
              source.path = cell.document.uri.toString();
              source.name = path.basename(cell.document.uri.path);
            }
          }
        });
      }
    };
  }
}

function hookSource(message: DebugProtocol.ProtocolMessage, hook: (source: DebugProtocol.Source | undefined) => void) {
  switch (message.type) {
    case "request":
      const req = message as DebugProtocol.Request;
      switch (req.command) {
        case "setBreakpoints":
          const sbr = req as DebugProtocol.SetBreakpointsRequest;
          hook(sbr.arguments.source);
          break;
        case "source":
          const sr = req as DebugProtocol.SourceRequest;
          hook(sr.arguments.source);
          break;
      }
      break;
    case "event":
      const evt = message as DebugProtocol.Event;
      switch (evt.event) {
        case "loadedSource":
          const lse = evt as DebugProtocol.LoadedSourceEvent;
          hook(lse.body.source);
          break;
      }
      break;
    case "response":
      const res = message as DebugProtocol.Response;
      switch (res.command) {
        case "stackTrace":
          const str = res as DebugProtocol.StackTraceResponse;
          str.body.stackFrames.forEach((frame) => {
            hook(frame.source);
          });
          break;
      }
      break;
  }
}
